### PR TITLE
Permalink variables

### PR DESCRIPTION
**Pull request by petzlux**
adding missing permalink variables:

- [x] Background Layer
- [x] Layers
- [x] Layer Opacity
- [x] Language
- [x] Location information (with crosshair) see https://github.com/Geoportail-Luxembourg/geoportailv3/commit/f0ec17e5c3ab05858a072b61c4a2657dfc6451f7

### CODE REVIEW

**Comment by petzlux** [comment]
please review @elemoine

**Comment by petzlux** [comment]
fixes #451

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                this.backgroundLayerMgr_.set(this.map_, layer);
+              }
+            }, this));
+
+        this.appThemes_.getThemesObject().then(
+            goog.bind(function(themes) {
+              if (goog.isDef(layerIds)) {
+                this.applyLayerStateToMap_(layerIds, themes);
+              }
+              this.setupWatchers_(selectedLayers);
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: Do you set up new watchers each time themes are loaded?

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+};
+
+
+/**
+ * @param {Array.<ol.layer.Layer>} layers
+ * @private
+ */
+app.LayerPermalinkManager.prototype.setLayerState_ = function(layers) {
+  var layerIds = goog.array.map(layers, function(layer) {
+    return layer.get('queryable_id');
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: We still don't have ids for layers? The only ids we have are the `queryable_id`s?

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+  goog.array.forEach(layerIds.reverse(),
+      function(layerId, layerIndex) {
+        var node = goog.array.find(flatCatalogue, function(catItem) {
+          return catItem.id === layerId;
+        });
+        var layer = this.getLayerFunc_(node);
+        // set opacity trough metadata to not interfere
+        // with the layer opacity manager service
+        layer.get('metadata')['start_opacity'] =
+            this.getStateValue_('opacities').reverse()[layerIndex];
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: Do you need to call `getStateValue_('opacitites')` for each layer id?

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+/**
+ * @param {string} parameter
+ * @private
+ * @return {Array.<number>|undefined}
+ */
+app.LayerPermalinkManager.prototype.getStateValue_ = function(parameter) {
+  var result = '';
+  var response = this.stateManager_.getInitialValue(parameter);
+  if (goog.isDef(response) && response.length > 0) {
+    result = this.stateManager_.getInitialValue(parameter);
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: Why do you call `getInitialValue` again?

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 3 above)
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: Also, it is strange to call `reverse()` on something that may be `undefined`. (The return type of `getStateValue_` is `Array.<number>|undefined`.)

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    var value = parseFloat(string);
+    if (goog.isNumber(value)) {
+      items.push(value);
+    }
+  });
+  if (items.length === 0) {
+    return undefined;
+  } else {
+    return items;
+  }
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: You could use the ternary notation here:

```js
  return items.length === 0 ? undefined : items;
```

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+  // Add event listeners to existing selected layers
+  goog.array.forEach(layers.getArray(), function(layer) {
+    goog.events.listen(layer, ol.ObjectEventType.PROPERTYCHANGE,
+        goog.bind(function() {
+          this.setLayerState_(selectedLayers);
+        }, this));
+  }, this);
+  // Add event listener to all future selected layers
+  goog.events.listen(layers, ol.CollectionEventType.ADD,
+      goog.bind(
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: You don't need to use `goog.bind` here as you pass `this` to `goog.events.listen` as the fifth argument (as done elsewhere).

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+  this.getLayerFunc_ = appGetLayerForCatalogNode;
+};
+
+
+/**
+ * @param {Array.<ol.layer.Layer>} layers
+ * @private
+ */
+app.LayerPermalinkManager.prototype.setLayerState_ = function(layers) {
+  var layerIds = goog.array.map(layers, function(layer) {
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: `goog.require('goog.array')`

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+            /**
+             * @param {Array.<ol.layer.Base>} bgLayers
+             */
+            function(bgLayers) {
+              var stateLayerLabel =
+                  this.stateManager_.getInitialValue('bgLayer');
+              if (goog.isDefAndNotNull(stateLayerLabel)) {
+                var layer = /** @type {ol.layer.Base} */
+                    (goog.array.find(bgLayers, function(layer) {
+                  return layer.get('label') === stateLayerLabel;
```
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: Isn't `label` a translated string? (not sure about this one)

File: geoportailv3/static/js/maincontroller.js [hunk 9]
```diff
@@ -269,6 +289,40 @@ app.MainController.prototype.switchLanguage = function(lang) {
 
 
 /**
+ * @private
+ */
+app.MainController.prototype.initLanguage_ = function() {
+  this.scope_.$watchCollection(goog.bind(function() {
```
Comment (elemoine) [inline geoportailv3/static/js/maincontroller.js]: `$watchCollection` for a scalar value? Are you sure?

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 1 above)
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: Remember that themes are reloaded when the user logs in.

File: geoportailv3/static/js/maincontroller.js (hunk 9 above)
Comment (petzlux) [inline geoportailv3/static/js/maincontroller.js]: `$watch` probably better ?

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 1 above)
Comment (petzlux) [inline geoportailv3/static/js/layerpermalinkservice.js]: `listenOnce` would help here ?

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 2 above)
Comment (petzlux) [inline geoportailv3/static/js/layerpermalinkservice.js]: AFAIK yes ...

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 1 above)
Comment (elemoine) [inline geoportailv3/static/js/layerpermalinkservice.js]: If you use `listenOnce` it won't work in the case where the user loads the application and then logs in, as you won't see the second `themes` response.

File: geoportailv3/static/js/maincontroller.js (hunk 9 above)
Comment (elemoine) [inline geoportailv3/static/js/maincontroller.js]: Does `$watchCollection` work at all in this case?

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 1 above)
Comment (petzlux) [inline geoportailv3/static/js/layerpermalinkservice.js]: Okay, I think i will have to use the de registration function the watcher returns to destroy it each time themes get reloaded ...

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 7 above)
Comment (petzlux) [inline geoportailv3/static/js/layerpermalinkservice.js]: so is the logic that we also need to add `goog.events` , `goog.bind`  ... to require statements ?

File: geoportailv3/static/js/layerpermalinkservice.js [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+  goog.events.listen(this.backgroundLayerMgr_, ngeo.BackgroundEventType.CHANGE,
+      function() {
+        var bgLayer = this.backgroundLayerMgr_.get(this.map_);
+        this.stateManager_.updateState({
+          'bgLayer': bgLayer.get('label')
+        });
+      }, undefined, this);
+
+  if (goog.isFunction(this.scopeWatcher_)) {
+    this.scopeWatcher_(); // destroy previous watcher
```
Comment (petzlux) [inline geoportailv3/static/js/layerpermalinkservice.js]: @elemoine is this better now?

File: geoportailv3/static/js/layerpermalinkservice.js (hunk 8 above)
Comment (petzlux) [inline geoportailv3/static/js/layerpermalinkservice.js]: Yes but it doesnt get translated internally, just for display in the background layer switcher, which I dont touch ... so no problem

**Comment by petzlux** [comment]
@elemoine  ready to merge now ?

**Pull request merged**